brainwidemap/encoding/pipelines/01_cache_regressors.py

```diff
@@ -0,0 +1,229 @@
+# Standard library
+import logging
+import os
+import pickle
+from datetime import datetime as dt
+from pathlib import Path
+
+# Third party libraries
+import dask
+import numpy as np
+import pandas as pd
+from dask.distributed import Client
+from dask_jobqueue import SLURMCluster
+
+# IBL libraries
+import brainbox.io.one as bbone
+from iblutil.util import Bunch
+from one.api import ONE
+from brainwidemap.encoding.params import GLM_CACHE
+from brainwidemap.encoding.utils import load_trials_df
+from brainwidemap.bwm_loading import load_trials_and_mask, bwm_query
+
+_logger = logging.getLogger("brainwide")
+
+
+def load_regressors(
+    session_id,
+    pid,
+    t_before=0.0,
+    t_after=0.0,
+    binwidth=0.02,
+    abswheel=False,
+    clu_criteria="bwm",
+    one=None,
+):
+    """
+    Load in regressors for given session and probe. Returns a dictionary with the following keys:
+
+    Parameters
+    ----------
+    session_id : str
+        EID of the session to load
+    pid : str
+        Probe ID to load associated with the session
+    t_before : float, optional
+        Time before stimulus onset to include in output trial_start column of df, by default 0.
+    t_after : float, optional
+        Time after feedback to include in output trial_end column of df, by default 0.
+    binwidth : float, optional
+        Binwidth for wheel signal. Needs to match that of GLM, by default 0.02
+    abswheel : bool, optional
+        Load in wheel speed instead of velocity, by default False
+    ret_qc : bool, optional
+        Whether to recompute cluster metrics and return a full dataframe of the result,
+        by default False
+    clu_criteria : str, optional
+        Criteria for saving clusters, 'all' for all units, 'bwm' for criteria matching that of
+        brain-wide map (all metrics passing). No others supported for now., by default 'bwm'
+    one : ONE, optional
+        Instance of ONE, by default None
+
+    Returns
+    -------
+    trialsdf, spk_times, spk_clu, clu_regions, clu_qc, clu_df, clu_qc (optional)
+        Output regressors for GLM
+    """
+    one = ONE() if one is None else one
```

Comment:
Usually would also prefer this but however for this script that causes problems. It's because this function is called within a Dask worker, and all function arguments must be sent to workers via a message-passing system. An ONE instance is rather cumbersome (several MB because of the cache) and when you have many workers it creates huge overhead on the dask messaging network.

I'm looking at another way of doing this that is more elegant but it will have to stay this way just for the moment.

Reply:
Ignore the above I realized a good way to elegantize it.

Reply:
Function `delayed_loadsave` in the cache script now instantiates ONE and passes it to this func, which has a mandatory positional argument for ONE